firebase-vertexai/src/main/kotlin/com/google/firebase/vertexai/type/LiveSession.kt

```diff
@@ -0,0 +1,358 @@
+
+  private fun fillRecordedAudioQueue() {
+    CoroutineScope(Dispatchers.IO).launch {
+      audioHelper!!.startRecording().collect {
+        if (!isRecording) {
+          cancel()
+        }
+        audioQueue.add(it)
+      }
+    }
+  }
+
+  private fun sendAudioDataToServer() {
+    CoroutineScope(Dispatchers.Default).launch {
+      val minBufferSize =
+        AudioTrack.getMinBufferSize(
+          24000,
+          AudioFormat.CHANNEL_OUT_MONO,
+          AudioFormat.ENCODING_PCM_16BIT
+        )
+      var bytesRead = 0
+      var recordedData = ByteArray(minBufferSize * 2)
+      while (true) {
+        if (!isRecording) {
+          break
+        }
+        val byteArr = audioQueue.poll()
+        if (byteArr != null) {
+          bytesRead += byteArr.size
+          recordedData += byteArr
+          if (bytesRead >= minBufferSize) {
+            sendMediaStream(listOf(MediaData("audio/pcm", recordedData)))
+            bytesRead = 0
+            recordedData = byteArrayOf()
+          }
+        } else {
+          continue
+        }
+      }
+    }
+  }
+
+  private fun fillServerResponseAudioQueue() {
+    CoroutineScope(Dispatchers.Default).launch {
+      receive(listOf(ContentModality.AUDIO)).collect {
+        if (!isRecording) {
+          cancel()
+        }
+        if (it.status == LiveContentResponse.Status.INTERRUPTED) {
+          while (!playBackQueue.isEmpty()) playBackQueue.poll()
+        } else if (it.status == LiveContentResponse.Status.NORMAL) {
+          if (!it.functionCalls.isNullOrEmpty()) {
+            functionCallChannel.send(it.functionCalls)
+          } else {
+            playBackQueue.add(it.data!!.parts[0].asInlineDataPartOrNull()!!.inlineData)
+          }
+        }
+      }
+    }
+  }
+
+  private fun playServerResponseAudio() {
+    CoroutineScope(Dispatchers.IO).launch {
+      while (true) {
+        if (!isRecording) {
+          break
+        }
+        val x = playBackQueue.poll()
+        if (x != null) {
+          audioHelper!!.playAudio(x)
+        }
+      }
+    }
+  }
+
+  /**
+   * Starts an audio conversation with the Gemini server, which can only be stopped using
+   * stopAudioConversation.
+   */
+  public suspend fun startAudioConversation() {
+    if (isRecording) {
+      return
+    }
+    functionCallChannel = Channel()
+    isRecording = true
+    audioHelper = AudioHelper()
+    audioHelper!!.setupAudioTrack()
+    val scope = CoroutineScope(Dispatchers.Default)
+    fillRecordedAudioQueue()
+    sendAudioDataToServer()
+    fillServerResponseAudioQueue()
+    playServerResponseAudio()
+    delay(1000)
+  }
+
+  /** Stops the audio conversation with the Gemini Server. */
+  public fun stopAudioConversation() {
+    stopReceiving()
+    isRecording = false
+    if (audioHelper != null) {
+      while (!playBackQueue.isEmpty()) playBackQueue.poll()
+      while (!audioQueue.isEmpty()) audioQueue.poll()
+      audioHelper!!.release()
+      audioHelper = null
+    }
```

Comment:
Anytime you use `!!` you ideally wanna think if there's another way to do what you're trying to do that avoids it.
You can also use `IsNotEmpty()` here.
```suggestion
    audioHelper?.let {
      while (playBackQueue.isNotEmpty()) playBackQueue.poll()
      while (audioQueue.isNotEmpty()) audioQueue.poll()
      it.release()
    }
    audioHelper = null
```